### PR TITLE
Shuffleboard PID tuning widgets

### PR DESCRIPTION
**Pull request by FireShredder24**
Add Shuffleboard PID widgets for drivetrain and shooter
Add more status widgets for shooter hood, shooter, and drivetrain

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -60,33 +60,36 @@
   private final ControllerBase mainController = new ControllerBase(xbox, f310);
 
   // ----------------------------------------Buttons------------------------------------------------
-  private final JoystickButton intakeRollerFwdButton = new JoystickButton(f310, LogitechF310.Button.kBumperLeft.value);
-  private final JoystickButton intakeRollerRvsButton = new JoystickButton(f310, LogitechF310.Button.kBumperRight.value);
-  private final JoystickButton intakePistonToggle = new JoystickButton(f310, LogitechF310.Button.kB.value);
+  private final JoystickButton intakeRollerFwdButton = new JoystickButton(f310, LogitechF310.Button.kBumperLeft.value); // Left
```
Comment (terbos90803) [inline src/main/java/frc/robot/RobotContainer.java]: Does the comment add anything here?  Is kBumperLeft not readable?  Adding a comment also adds maintenance.

Same feedback for all the comments added here.

File: src/main/java/frc/robot/commands/DriveCommand.java [hunk 2]
```diff
@@ -35,7 +35,7 @@ public void initialize() {
   @Override
   public void execute() {
     m_subsystem.drive(m_conditioner.conditionMove(m_controller.getMoveRequest()),
-        m_conditioner.conditionTurn(m_controller.getTurnRequest()));
+        m_conditioner.conditionTurn(-m_controller.getTurnRequest()));
```
Comment (terbos90803) [inline src/main/java/frc/robot/commands/DriveCommand.java]: Why was this needed now?  It was working correctly before.

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java [hunk 3]
```diff
@@ -58,31 +59,44 @@
...
    */
-  private static final double K_FF = 1;
+  private double kFF = 1;
 
   private final DifferentialDriveKinematics kinematics = new DifferentialDriveKinematics(
       Units.inchesToMeters(DRIVE_BASE_WIDTH_INCHES));
 
+  private NetworkTableEntry leftDriveVelocity;
+  private NetworkTableEntry rightDriveVelocity;
+  private NetworkTableEntry kPSlider;
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java]: There used to be a combined element to show all the PID coeffs together.  Is that still around?

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java [hunk 4]
```diff
@@ -103,7 +109,8 @@ private boolean hasReliableZero() {
   }
 
   private boolean isFullyDeployed() {
-    return Math.abs(getHoodPos() - currSetpoint) < MoPrefs.getShooterHoodPositionTolerance();
+    return MoPrefs.getShooterHoodSetpoint() - MoPrefs.getShooterHoodPositionTolerance() < getHoodPos()
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java]: Making tolerance comparisons should always be in the form:
Math.abs(setpoint - currPos) < tolerance

Why did this get changed?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 5]
```diff
@@ -165,19 +198,38 @@ public void purge() {
...
   }
 
   private boolean isFlywheelReady() {
-    return Math.abs(MoPrefs.getShooterPIDSetpoint() - shooterEncoder.getVelocity()) < MoPrefs
-        .getShooterFlywheelTolerance();
+    return Math.abs(setpoint - shooterEncoder.getVelocity()) < MoPrefs.getShooterFlywheelTolerance();
+  }
+
+  private void updatePIDConstants() {
+    shooterPIDLeft.setP(kP, 0);
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: You can't have two motors mechanically connected and also PID them separately.  The PID has to be on one of the motors and the other motor follows the first one.

**Review by terbos90803 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/DriveCommand.java (hunk 2 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/commands/DriveCommand.java]: Well, turning was inverted.  I don't know why, but it was.  I'd guess that some tweaking to Conditioners over the summer did that.

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java (hunk 3 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java]: There is, but it doesn't work with Talon FX(or Spark MAX), since neither extends the WPILib PIDController class.

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java (hunk 4 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java]: I was having a lot of trouble with the it, so I messed around with the logic.  Eventually, it turned out that getHoodPos() wasn't actually polling the encoder.  I just didn't change it back because we moved on to tuning the shooter flywheel PID.  I'll fix it once I finish my homework.

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/RobotContainer.java]: In hindsight, it didn't help much.  Will revert after homework is done.

File: src/main/java/frc/robot/commands/DriveCommand.java (hunk 2 above)
Comment (terbos90803) [reply src/main/java/frc/robot/commands/DriveCommand.java]: It's much more important to understand why than to just patch until working.  Doing that often leads to a big mess.
For instance, it is normal to invert the raw Y axis because of the long history of video games on legacy computers where Y is positive down on the screen.
It is not normal to invert X, so I'd look into this a lot more closely.

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java (hunk 3 above)
Comment (terbos90803) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java]: I thought Jordan wrote a wrapper that didn't rely on that class.  Or am I mis-remembering.

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java (hunk 4 above)
Comment (terbos90803) [reply src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java]: Another case of needing better understanding before hacking.
I get that if you aren't familiar with the standard tolerance pattern that it may have seemed odd, but now it seems odd and difficult to read for the rest of us.

As a general rule, any time you have to work to understand a bit of code, that's a screaming flag that it deserves a comment block to explain what's going on and why.  This is doubly true if you need to change something to work around a bug.  When the real bug gets fixed, it needs to be easy to find any workarounds that will no longer behave as expected and fix them too.

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java (hunk 3 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java]: According to the Java API, the widget kPIDController ([here](https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/shuffleboard/BuiltInWidgets.html#kPIDController)) relies on a deprecated version of PIDController.  It's weird.  For now, I've just been using separate widgets, organizing them into lists.  It's not very screen-space-efficient, but it works.

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java [hunk 6]
```diff
@@ -91,19 +98,24 @@ public void stowHood() {
   }
 
   public double getHoodPos() {
-    return hoodPos;
+    return hoodEncoder.getPosition();
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java]: Just out of curiosity, why did this move here?  It used to be in periodic().  Unless we truly need to get the hood position more frequently than regular period, it's better to read it once and cache the result.

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java [hunk 7]
```diff
@@ -112,21 +124,23 @@ public boolean isHoodReady() {
 
   @Override
   public void periodic() {
-    hoodPos = hoodEncoder.getPosition();
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:115]: This line needs to come back.

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 8]
```diff
@@ -125,16 +126,46 @@ public ShooterSubsystem(final ShooterHoodSubsystem shooterHood, ShuffleboardTab
...
+
+    // Adds a Shuffleboard widget to show whether the flywheel is spinning within a
+    // certain tolerance of the setpoint. See isFlywheelReady().
+    isFlywheelReady = tab.add("Is Flywheel Ready?", false).withWidget(BuiltInWidgets.kBooleanBox).getEntry();
+
+    shooterSetpoint = tab.addPersistent("Shooter Setpoint", 4500).withWidget(BuiltInWidgets.kTextView).getEntry();
+    shooterSetpoint.addListener(notice -> setpoint = notice.value.getDouble(),
+        EntryListenerFlags.kNew | EntryListenerFlags.kUpdate);
+
+    kPSlider = tab.addPersistent("Shooter kP", 5e-5).withWidget(BuiltInWidgets.kTextView).getEntry();
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: There should be no hardcoded constants in these calls.  Especially since these are tuning values and already exist elsewhere.  Keeping them in sync is high maintenance.

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 9]
```diff
@@ -125,16 +126,46 @@ public ShooterSubsystem(final ShooterHoodSubsystem shooterHood, ShuffleboardTab
...
+    kFFSlider.addListener(notice -> kFF = notice.value.getDouble(),
+        EntryListenerFlags.kNew | EntryListenerFlags.kUpdate);
   }
 
   public void shoot(double hoodSetpoint) {
     // fast shooter wheel
     // deploy hood
     // run gate if both of "" are good
 
-    double pidSetpoint = MoPrefs.getShooterPIDSetpoint();
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:137]: What the heck is going on here?  The setpoint should be a MoPref and not hardcoded (aside from a default).

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 10]
```diff
@@ -165,19 +196,33 @@ public void purge() {
...
   @Override
   public void periodic() {
-    SmartDashboard.putNumber("Flywheel Speed", shooterEncoder.getVelocity());
-    SmartDashboard.putNumber("Flywheel Position", shooterEncoder.getPosition());
-    SmartDashboard.putBoolean("Flywheel ready?", isFlywheelReady());
+    flywheelSpeed.setDouble(shooterEncoder.getVelocity());
+    isFlywheelReady.setBoolean(isFlywheelReady());
+    updatePIDConstants();
+
+    setpoint = shooterSetpoint.getDouble(4500);
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: What's going on here?  Why is the setpoint being updated every periodic and there's a hardcoded constant which is repeated multiple times in the code.

**Review by terbos90803 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java (hunk 6 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java]: Fixed

File: src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java (hunk 7 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:115]: It has

File: src/main/java/frc/robot/commands/DriveCommand.java (hunk 2 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/commands/DriveCommand.java]: Jordan fixed this by adding a FunctionalConditioner to the ComposedConditioner that is passed to DriveCommand in RobotContainer.  It reverses turnRequest while leaving moveRequest untouched.

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java [hunk 11]
```diff
@@ -185,4 +159,40 @@ public void stop() {
...
+  }
+
+  // Updates the PID constants stored in each Talon FX from MoPrefs
+  private void updatePIDConstants() {
+    double kP = MoPrefs.getDriveKP();
+    double kI = MoPrefs.getDriveKI();
+    double kD = MoPrefs.getDriveKD();
+    int kIZ = (int) MoPrefs.getDriveKIZ();
+    double kFF = MoPrefs.getDriveKFF();
+    leftFront.config_kP(0, kP);
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:181]: I know it adds complexity, but I seem to recall that you shouldn't be writing PID constants every period.  You'll have to figure out how to only write them on changes.  That usually requires caching the old value and only writing on diffs.  Jordan can probably help make this look all fancy so the code doesn't become cumbersome.

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 12]
```diff
@@ -165,19 +139,32 @@ public void purge() {
...
+    return Utils.map(closedLoopSetpoint, -MAX_FREE_SPEED, MAX_FREE_SPEED, -1, 1);
   }
 
   private boolean isFlywheelReady() {
     return Math.abs(MoPrefs.getShooterPIDSetpoint() - shooterEncoder.getVelocity()) < MoPrefs
         .getShooterFlywheelTolerance();
   }
 
+  private void updatePIDConstants() {
+    shooterPIDRight.setP(MoPrefs.getShooterKP(), 0);
```
Comment (terbos90803) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: Only update on change

File: src/main/java/frc/robot/utils/MoPrefs.java [hunk 13]
```diff
@@ -86,4 +109,72 @@ public static double getShooterPIDSetpoint() {
   public static double getShootFromWallHoodSetpoint() {
     return getDouble("Shoot From Wall Hood Setpoint", SHOOT_FROM_WALL_HOOD_SETPOINT);
   }
+
+  public static double getShooterKP() {
```
Comment (terbos90803) [inline src/main/java/frc/robot/utils/MoPrefs.java]: Does anyone else find this code pattern a challenge to maintain?
Why do we need specific methods baked into a shared class to get values from a dictionary?
It has always seemed to me that this should be easier.  Specifically, MoPrefs should maintain the dictionary, but not know/care about what's in it.  The clients should be responsible for knowing their own keys.

**Review by terbos90803 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/utils/MoPrefs.java (hunk 13 above)
Comment (jordan-powers) [reply src/main/java/frc/robot/utils/MoPrefs.java]: Well, MoPrefs is just a wrapper around ` edu.wpi.first.wpilibj.Preferences`. The only functionality it adds is the individual method calls. As far as I recall, these are the reasons for why we chose to create the MoPrefs instead of directly using the Preferences object:
1. To provide explicit default values for each preference
2. To make it very obvious what values are stored in the preferences
3. To prevent errors where typos in a string key ends up getting unexpected values

However, I do agree that this way of achieving that goal is pretty clunky.

Maybe a better way to do this would be to create a `enum MoPrefKeys` and refactor `MoPrefs` to only have one public method which takes a MoPrefKey and returns the value. Java's enums are basically just classes, so we could define a MoPrefKey to hold both the string key and an explicit default value.

That shouldn't be too hard to implement: I'll create a PR.

Comment (jordan-powers) [reply src/main/java/frc/robot/utils/MoPrefs.java]: See #27

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java (hunk 11 above)
Comment (jordan-powers) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:181]: This would be easy if the Preferences class allowed us access to the underlying `NetworkTableEntry`s.  Unfortunately, that is not the case.
Also, references is a final class, so we can't subclass it and give ourselves access.

I can see two options:
1. Re-implement the Preferences class, and add a method to get access to the `NetworkTableEntry`. The Preferences class is pretty simple, so this wouldn't be too difficult. It does seem a bit hacky though.
2. Create some sort of "watcher" class that watches a value, looks for changes, and triggers a lambda when the value changes.  This would be pretty trivial to implement, however it would require polling the Preferences NetworkTable on every loop. This is probably not a big deal, though, and I suspect the NetworkTables are designed to support it.

I'm not sure which option is better.

Comment (terbos90803) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:181]: I thought MoPrefs already subclassed Preferences.  Apparently, it does not, so it must wrap it instead (composition vs inheritance).

I also thought there was a way to subscribe to NT entries, but if there's not, then having a mechanism in place to detect and notify on changes would certainly be helpful.

File: src/main/java/frc/robot/utils/MoPrefs.java [hunk 14]
```diff
@@ -47,43 +123,11 @@ public static void setDouble(String key, double value) {
     System.out.format("set pref: %s=%f\n", key, value);
```
Comment (aimiebell) [inline src/main/java/frc/robot/utils/MoPrefs.java:123]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.

File: src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java (hunk 11 above)
Comment (FireShredder24) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:181]: What if instead of polling Preferences and writing the values to the Talon FXs every scheduler cycle, we only did that once per second?  That would be trivially easy to implement and would basically accomplish the same goal.

Comment (terbos90803) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:181]: It might help to understand the root problem.
The SparkMax docs say specifically that certain values like the PID settings shouldn't be written frequently.  I suspect the TalonFX has a similar warning.
The reason is because those values are held in a form of non-volatile memory (flash or eeprom) which has a limited number of write cycles in its lifetime.  You want to make sure you only write that when you're changing it.  Writing the same value repeatedly only wears out the storage and shortens the life of the device.

Theoretically, this type of check should be happening in the vendor code between you and the device, but you can never know for sure and if the docs say don't do it, then it's a safer bet to assume they aren't checking for you.

Comment (jordan-powers) [reply src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:181]: There is a way to subscribe to Networktable entries. However, the WPILib Preferences class doesn't let us access those entries. Just the values.
